c++/src/Compression.cc

```diff
@@ -564,16 +591,47 @@ DIAGNOSTIC_POP
     return bytesReturned;
   }
 
+  /** There are three possible scenarios when seeking a position:
+   * 1. The seeked position is already read and decompressed into
+   *    the output stream.
+   * 2. It is already read from the input stream, but has not been
+   *    decompressed yet, ie. it's not in the output stream.
+   * 3. It is not read yet from the inputstream.
+   */
   void ZlibDecompressionStream::seek(PositionProvider& position) {
-    // clear state to force seek to read from the right position
+    size_t seekedPosition = position.current();
+    // Case 3.: the seeked position is the one that is currently buffered and
+    // decompressed. Here we only need to set the output buffer's pointer to the
+    // seeked position. Since headerPos saves the position after reading the
+    // first 3 bytes of the header, it is compared to the looked up pos + 3.
+    if (headerPosition == seekedPosition + 3
+        && inputBufferStartPosition <= headerPosition && inputBufferStart) {
+      position.next(); // Skip the input level position.
+      size_t posInChunk = position.next(); // Chunk level position.
+      outputBufferLength = uncompressedBufferLength - posInChunk;
+      outputBuffer = outputBufferStart + posInChunk;
+      return;
+    }
+    // Clear state to prepare reading from a new chunk header.
     state = DECOMPRESS_HEADER;
     outputBuffer = nullptr;
     outputBufferLength = 0;
     remainingLength = 0;
-    inputBuffer = nullptr;
-    inputBufferEnd = nullptr;
+    if (seekedPosition < static_cast<uint64_t>(input->ByteCount())) {
+      // Case 2.: The input is buffered, but not yet decompressed. No need to
+      // force re-reading the inputBuffer, we just have to move it to the
+      // seeked position.
+      position.next(); // Skip the input level position.
+      inputBuffer
+          = inputBufferStart + (seekedPosition - inputBufferStartPosition);
```

Comment:
If seekedPosition is smaller than inputBufferStartPosition, the inputBufferStart does not contain the data we need. This can happen when users call RowReader::seekToRow() in a reverse order.

Reply:
Added condition.